### PR TITLE
Feat: manage api  support route group

### PR DESCRIPTION
**Pull request by liuxiran**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [ ] Bug fix
- [x] New feature provided
- [ ] Improve performance

- Related issues
fix issue: #431 
___
### Bugfix
- Description

- How to fix?

___
### New feature or improvement
- Describe the details and related test reports.

- add a new module routeGroup: 

  -  create table `route_group`

  - support CRUD and list route group

- create/update route support add attribute route group id

### CODE REVIEW

File: api/route/route.go [hunk 1]
```diff
@@ -101,11 +101,15 @@ func listRoute(c *gin.Context) {
 		db = db.Where("upstream_nodes like ? ", "%"+ip+"%")
 		isSearch = false
 	}
+	if rgid, exist := c.GetQuery("route_group_id"); exist {
+		db = db.Where("route_group_id equal ?", rgid)
```
Comment (nic-chen) [inline api/route/route.go:105]: does `equal` works  ?

File: api/route/route_group.go [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+		e := errno.FromMessage(errno.RouteGroupRequestError, err.Error())
+		logger.Error(e.Msg)
+		c.AbortWithStatusJSON(http.StatusInternalServerError, e.Response())
+		return
+	} else {
+		result := &service.ListResponse{Count: count, Data: routeGroupList}
+		resp, _ := json.Marshal(result)
+		c.Data(http.StatusOK, service.ContentType, resp)
+	}
+}
```
Comment (nic-chen) [inline api/route/route_group.go:105]: need a blank line between two functions.

File: api/route/route_group.go [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+func listRouteGroupName(c *gin.Context) {
+	db := conf.DB()
```
Comment (nic-chen) [inline api/route/route_group.go:107]: i think we should make paging for this api.

File: api/service/route_group.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+	}
+}
+
+func (rd *RouteGroupDao) UpdateRouteGroup() error {
+	db := conf.DB()
+	return db.Model(&RouteGroupDao{}).Update(rd).Error
+}
+
+func (rd *RouteGroupDao) DeleteRouteGroup() error {
+	return conf.DB().Delete(&rd).Error
```
Comment (nic-chen) [inline api/service/route_group.go:73]: we should check if there is a route bound before delete

**Review by nic-chen — CHANGES_REQUESTED** [review]

File: api/service/route_group.go (hunk 4 above)
Comment (liuxiran) [reply api/service/route_group.go:73]: yeap，and it has been checked in `function deleteRouteGroup` in `api/route/route_group.go ` file

Comment (liuxiran) [reply api/service/route_group.go:84]: this check has been put in the RoutegroupDao before, and tansfered to the service after updated :)

File: api/route/route_group.go (hunk 3 above)
Comment (liuxiran) [reply api/route/route_group.go:108]: this is used in select routegroup name in selectbox, and I found that the same method in upstream has not made paging too. so would it need to modified at the same time? @nic-chen

Comment (nic-chen) [reply api/route/route_group.go:142]: yes if you have time to do that :)

**Comment by juzhiyuan** [comment]
@nic-chen any update?

File: api/route/route_group.go (hunk 3 above)
Comment (liuxiran) [reply api/route/route_group.go:142]: ok, I will open other pr to make paging about  these two apis :)

File: api/route/route_group.go (hunk 2 above)
Comment (liuxiran) [reply api/route/route_group.go:139]: done

File: api/route/route.go (hunk 1 above)
Comment (liuxiran) [reply api/route/route.go]: done

File: api/route/route.go [hunk 5]
```diff
@@ -209,6 +223,15 @@ func updateRoute(c *gin.Context) {
 			}
 		}()
 		logger.Info(rd)
+		if isCreateGroup {
+			if err := tx.Model(&service.RouteGroupDao{}).Create(routeGroup).Error; err != nil {
+				tx.Rollback()
```
Comment (nic-chen) [inline api/route/route.go:228]: why we need a transaction here ?

Comment (juzhiyuan) [reply api/route/route.go:228]: @liuxiran Please take look :D

Comment (liuxiran) [reply api/route/route.go:228]: Refer to the original design, route must be in a group, if the group has  not exited, create with the route. so here if the group create with error, do rollback :) @nic-chen

Comment (nic-chen) [reply api/route/route.go:228]: @liuxiran  Sorry, it looks like a separate transaction.

**Review by nic-chen — APPROVED** [review]

**Pull request merged**